app/controllers/AddressLookupController.scala

```diff
@@ -79,40 +80,58 @@ class AddressLookupController @Inject()(journeyRepository: JourneyRepository, ad
     }
   }
 
+  sealed trait ResultsCount
+  case class OneResult(res: ProposedAddress) extends ResultsCount
+  case class ResultsList(res: Seq[ProposedAddress]) extends ResultsCount
+  case object NoResults extends ResultsCount
+  case object TooManyResults extends ResultsCount
+
   // GET  /:id/select
   // show a list of proposals from lookup parameters; always do the remote lookup as the parameters may have changed
   // go back to the lookup form on form binding error
   // we could optimize this to check whether or not parameters have changed but not really worth the effort at present
   def select(id: String) = Action.async { implicit req =>
     withFutureJourney(id) { journeyData =>
-      lookupForm.bindFromRequest().fold(
+      lookupForm.bindFromRequest() fold (
         errors => Future.successful((None, BadRequest(views.html.lookup(id, journeyData, errors)))),
-        lookup => {
-          addressService.find(lookup.postcode, lookup.filter).map {
-            case noneFound if noneFound.isEmpty => (None, Ok(views.html.lookup(id, journeyData, lookupForm.fill(lookup), Some(journeyData.config.lookupPage.getOrElse(LookupPage()).noResultsFoundMessage.getOrElse("Sorry, we couldn't find anything for that postcode.")))))
-            case oneFound if oneFound.size == 1 => (Some(journeyData.copy(selectedAddress = Some(oneFound.head.toConfirmableAddress(id)))), Redirect(routes.AddressLookupController.confirm(id)))
-            case tooManyFound if tooManyFound.size > journeyData.config.selectPage.getOrElse(SelectPage()).proposalListLimit.getOrElse(tooManyFound.size) =>
-              (None, Ok(views.html.lookup(id, journeyData, lookupForm.fill(lookup), Some(journeyData.config.lookupPage.getOrElse(LookupPage()).resultLimitExceededMessage.getOrElse("There were too many results. Please add additional details to limit the number of results.")))))
-            case displayProposals => (Some(journeyData.copy(proposals = Some(displayProposals))), Ok(views.html.select(id, journeyData, selectForm, Proposals(Some(displayProposals)))))
-          }
+        lookup => handleLookup(id, journeyData, lookup) map {
```

Comment:
This is just indirection. There's no need to delegate `addressService.find` to a helper function and then match again on a case typed return from that for a second time. I don't understand why this change was necessary. Please restore.

Reply:
This is required because of Line 112, the handleLookup function is recursive due to a logical issue with the address filter being on the backend and the frontend requirements being to serve the results for the postcode when no exact match address was found for a particular filter.